src/main/scala/is/hail/expr/Type.scala

```diff
@@ -1469,6 +1472,115 @@ final case class TStruct(fields: IndexedSeq[Field]) extends Type {
     filter(fn)
   }
 
+  def ungroup(x: Array[String]): (TStruct, (Row) => Row) = {
+    val newFieldNames = mutable.Set(fieldNames.toSet.diff(x.toSet).toArray: _*)
+
+    val structsToUngroup = x.map{ name =>
+      fieldOption(name) match {
+        case None => fatal(s"Struct does not have field with name `$name'.")
+        case Some(fd) =>
+          if (!fd.typ.isInstanceOf[TStruct])
+            fatal(s"Expecting a type of TStruct for field `$name', but found ${ fd.typ.toPrettyString() } ")
+
+          val s = fd.typ.asInstanceOf[TStruct]
+          val fieldsToAdd = s.fields.map(_.name).toArray
+
+          val dupFieldNames = fieldsToAdd.toSet.intersect(newFieldNames)
+          if (dupFieldNames.nonEmpty)
+            fatal(s"Found fields in struct `$name' has names overlapping with existing field names: ${ dupFieldNames.mkString(", ") }.")
+
+          newFieldNames ++= fieldsToAdd
+
+          (fd.index, (s, query(name), s.fields.length))
+      }
+    }.toMap
+
+    val nFieldsInit = fields.length
+    val nFieldsToInsert = structsToUngroup.map { case (_, (_, _, n)) => n }.sum
+    val nNewFields = nFieldsInit + nFieldsToInsert - structsToUngroup.size
+
+    val newFields = Array.fill[Field](nNewFields)(null)
```

Comment:
I'm finding this approach somewhat hard to parse. There's interplay between i and localIdx, but it's not checked. 
A functional thing that does something like
```scala 
fields.flatMap(f => (the field or its subfields))
    .zipWithIndex
    .map { case (f, i) => f.copy(index = i) }
```
may be easier to understand